### DISCUSSION TITLE
Inconsistent getServerSideProps behaviour using NextJS in vercel

### DISCUSSION THREAD

**Discussion by dgrcode** [vercel/next.js#39337] [Help]
I have the following setup

 - `/template`. Loads a page that shows all the templates owned by the current user.
 - `/template/new`. Loads a page that allows the user to create a template and save it.
 - `/api/save-template`. Receives the frontend call made from `/template/new` and handles storing it on the db.

With those routes, a typical use case is the following:
1. User ends up somehow (irrelevant) on `/template/new`.
2. User creates a template and hit "save". A spiner shows up while saving happens.
3. Template payload is sent via POST HTTP request to `/api/save-template`.
4. `/api/save-template` route handles the saving successfully and responds with a status `200`.
5. The spinner finishes and the user is redirected to `/template`, where the newly created template should hopefully show up.

The problem here is on point 5, where in some scenarios, the newly created template doesn't show up. Here's what happens on different scenarios. The app could be running either using `next dev`, `next build && next start`, or deployed on vercel. The database could either be hosted locally on docker, or on the cloud using supabase.

| db \ app | `next dev` | `next build...` | vercel |
|---|---|---|---|
| docker| ✅ | ✅ | haven't tried|
| supabase | ✅ | ✅ | ❌ |

One thing to keep in mind, which might be related to the reason this is not working, is that the redirection is made on the frontend using `router.push('/template')`.

I'm also checking on the client side what templates are available when reaching point 5. I've confirmed this is not a db delay issue because the templates are found by the client. For those checks I'm using the following code:
```js
useEffect(() => {
    const checkSupabase = async () => {
      const templatesResponse = await supabase
        .from<Template[]>('Template')
        .select('*')
      const templates = templatesResponse.data
      console.log('templates found from useEffect')
      console.log(templates)
    }

    checkSupabase()
  }, [])
```

And the last thing I want to point out is that the logs I get from vercel make me think the `getServerSideProps` is not called. I'm logging `templates found from getServerSideProps` to see what's available on that function. I see the log before creating the template (time 21:31:24:00), and then the `/api/save-template` being hit (time 21:31:47:11), but I don't see the second request to `/template` after the redirect. The logs –note the order is newest on the top– are:

```
21:31:47:11 - [POST] /api/save-template


21:31:24:00 - [GET] /template
templates found from getServerSideProps
[]
```

## Questions
1. How is it that I don't see a second log for `[GET] /template` or the `templates found from getServerSideProps` log? Is it not being fired?
2. If `getServerSideProps` is not fired, how come this works when the app is running locally?
3. Is there a consistent solution for this use case? I guess it must be quite common to want to see recently created data.

Thanks! 

[Also asked on [stackoverflow](https://stackoverflow.com/questions/73241640/inconsistent-getserversideprops-behaviour-using-nextjs-in-vercel)]

**Comment by wiiiimm** [comment]
Hi,

I have deployed a simple SSR example over at https://93077-nextjs-get-server-side-props.vercel-support.app/ (source code available at: https://github.com/vercel-support/93077-nextjs-getServerSideProps) and you can click around the internal links (created using `next/link`) to switch around a static page and a SSR page that display a timestamp obtained inside the `getServerSideProps` function from the time API at https://nextjs-time-api.vercel-support.app/api/time.

This example illustrates that `getServerSideProps` do rerender for every request. As for the problem you described, can you check to see if you have some `cache-control` header set for your pages so a cached result is used instead of one generated from SSR?

Personally, for use case like yours, I usually load my template records in the /template page directly on the client side instead of relying on SSR as it offers a better and quicker frontend experience.

**Reply by dgrcode** [reply]
I've checked the `cache-control` headers are set to `private, no-cache, no-store, max-age=0, must-revalidate`, which means things should work as expected. The problem is that the call to `/template` never happens, not that it happens but it's returned cached information.

Regarding your recommendation about loading in the client, I'm currently working on implementing a client side cache with react query. I'll hopefully deploy that today. Once I have it in place it will be fairly easy to remove the server side rendering.

In case it helps debugging what' going on, following are the steps I've done and what I've found in them.

### 1. Go to `/template`
I've checked the stuff I got from the network.
<img width="832" alt="image" src="https://user-images.githubusercontent.com/17493200/184619298-a0894618-f332-432a-ba93-962467a800d4.png">


I cleaned the database, so here I can see I don't have any template. All good so far.

One thing I didn't expect is that I'm also downloading the rest of data for the pages accessible from the current one. Those are all the jsons shown on the left hand side list. I remember reading that's for improved loading of the pages. I guess at some point I might want to turn that off if I hit the database too much, but for now that's not a problem.

Now I clean the network tab to see what comes on the next page load.

### 2. Go to `/template/new` (clicking on the ui link, although I guess this shouldn't matter)
<img width="788" alt="image" src="https://user-images.githubusercontent.com/17493200/184619532-cd0c6ea1-8f4c-4988-9514-d91e4dd60a30.png">


Nothing loads, which makes sense because this page doesn't have `getServerSideProps`. Also, any page reachable from here got the data dowloaded on the previos step.

Now I clear the network tab again, and proceed to create a new template with the ui.

### 3. Save the template and get redirected to `/template`
The redirect happens with `router.push('/template')`, in case that matters.
<img width="832" alt="image" src="https://user-images.githubusercontent.com/17493200/184619559-6ff1cce6-20d0-4fb1-ad7c-9fc0970ea2a3.png">

Here we can see how there's no call to `/template`. So it's not like the call was made and the cache was returned. The call never happened, which is what I don't understand.

When refreshing the page the result I get is similar to step 1, where the whole data is loaded again, and I get the expected results. But only after manually refreshing the page. It looks like client side navigation somehow skips getting fresh data from `getServerSideProps`.

**Reply by wiiiimm** [reply]
@dgrcode can you try to recreate this issue in a new clean Next.js repo and share the repo over here and I can take a look at your code?

**Reply by dgrcode** [reply]
Sure, I'm on this. I'll reply once I have a working example with the same behaviour. Hopefully very soon 🤞

**Reply by dgrcode** [reply]
I haven't been able to reproduce it yet. I created a very simple example, and it seems to be working as expected. I have to leave this for today, but I'll see if I can continue adding more complexity to the app so it gets closer in behaviour to what I have in my actual app.

In case you want to have a look (although things work as expected):
- repo: https://github.com/dgrcode/nextjs-getserversideprops-issue
- live app: https://nextjs-getserversideprops-issue.vercel.app/
